File: 2024/day_13/solution.py
def is_robot_back(moves: str) -> bool | list[int]:
  x, y = 0, 0
  visited = set()

  move_delta = {
    'L': (-1, 0),
    'R': (1, 0),
    'U': (0, 1),
    'D': (0, -1),
  }

  def invert_move(move: str) -> str:
    inverted = {
      'L': 'R',
      'R': 'L',
      'U': 'D',
      'D': 'U',
    }

    return inverted.get(move, move)

  i = 0
  while i < len(moves):
    move = moves[i]
    intensity = 1

    if move == '*':
      intensity = 2
      i += 1
      move = moves[i]
    elif move == '!':
      i += 1
      move = invert_move(moves[i])
    elif move == '?':
      i += 1
      next_move = moves[i]
      if next_move in visited:
        i += 1
        continue
      move = next_move

    if move in move_delta:
      dx, dy = move_delta[move]
      x += dx * intensity
      y += dy * intensity
      visited.add(move)
      
    i += 1

  return True if x == 0 and y == 0 else [x, y]

File: 2024/day_13/solution.js
function isRobotBack(moves) {
  let x = 0,
    y = 0;
  let inverted = false;
  let visited = new Set();

  const moveActions = {
    L: (intensity) => (x -= intensity),
    R: (intensity) => (x += intensity),
    U: (intensity) => (y += intensity),
    D: (intensity) => (y -= intensity),
  };

  const invertedMoves = {
    L: "R",
    R: "L",
    U: "D",
    D: "U",
  };

  for (let i = 0; i < moves.length; i++) {
    let move = moves[i];
    let intensity = 1;

    switch (move) {
      case "*":
        intensity = 2;
        move = moves[++i];
        break;
      case "!":
        inverted = !inverted;
        move = moves[++i];
        break;
      case "?":
        let nextMove = moves[++i];
        if (visited.has(nextMove)) continue;
        move = nextMove;
        break;
    }

    if (inverted) {
      move = invertedMoves[move] || move;
      inverted = false;
    }

    visited.add(move);

    if (moveActions[move]) {
      moveActions[move](intensity);
    }
  }

  return x === 0 && y === 0 ? true : [x, y];
}

console.log(isRobotBack("R")); // [1, 0]
console.log(isRobotBack("RL")); // true
console.log(isRobotBack("RLUD")); // true
console.log(isRobotBack("*RU")); // [2, 1]
console.log(isRobotBack("R*U")); // [1, 2]
console.log(isRobotBack("LLL!R")); // [-4, 0]
console.log(isRobotBack("R?R")); // [1, 0]
console.log(isRobotBack("U?D")); // true
console.log(isRobotBack("R!L")); // [2, 0]
console.log(isRobotBack("U!D")); // [0, 2]
console.log(isRobotBack("R?L")); // true
console.log(isRobotBack("U?U")); // [0, 1]
console.log(isRobotBack("*U?U")); // [0, 2]
console.log(isRobotBack("U?D?U")); // true
console.log(isRobotBack("R!U?U")); // [1, 0]
console.log(isRobotBack("UU!U?D")); // [0, 1]
